### PR TITLE
Redo all Discrete Transition nodes

### PR DESCRIPTION
**Pull request by wouterwln**
Now adds 1(!!!) function to govern all Discrete Transition categorical rules. Probably the generic implementation is a bit slow at times but I will work on performance (and maybe add a few alternative strategies for when we have a nice and known structure)

### CODE REVIEW

File: test/rules/discrete_transition/in_tests.jl [hunk 1]
```diff
@@ -4,13 +4,6 @@
 
     import ReactiveMP: @test_rules
 
-    @testset "Belief Propagation: (m_out::Categorical, m_a::PointMass)" begin
-        @test_rules [check_type_promotion = false] DiscreteTransition(:in, Marginalisation) [(
-            input = (m_out = Categorical([0.1, 0.4, 0.5]), m_a = PointMass([0.2 0.1 0.7; 0.4 0.3 0.3; 0.1 0.6 0.3])),
-            output = Categorical([0.23000000000000004, 0.43, 0.33999999999999997])
-        )]
-    end
-
```
Comment (bvdmitri) [inline test/rules/discrete_transition/in_tests.jl:13]: Were these wrong?

**Review by bvdmitri — COMMENTED** [review]
Looks neat, but some tests were either removed or changed. Fine with me, because those tests were checking some numerical answers? But what is the reason for changing tests in `in_tests.jl` and `out_tests.jl`?

File: test/rules/discrete_transition/in_tests.jl (hunk 1 above)
Comment (wouterwln) [reply test/rules/discrete_transition/in_tests.jl:13]: This test is wrong, since we cannot get an incoming message on `a`, we can only set it to `PointMass` and it will become a marginal (for which the rule is implemented). The rule for VMP was actually wrong, as was the test, and this was caught by the new general rule (I double checked this with Thijs)

**Comment by bvdmitri** [comment]
@wouterwln there is a failing test in `RxInfer` repo called "`infer` with UnfactorizedData"

This is the stacktrace 

```
ERROR: LoadError: MethodError: no method matching sum_out_dimensions(::Matrix{Float64}, ::Tuple{Int64}, ::Vector{Int64})
The function `sum_out_dimensions` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  sum_out_dimensions(::AbstractArray{T, M}, ::NTuple{N, Int64}, !Matched::AbstractArray{T, N}) where {T, M, N}
   @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:30

Stacktrace:
   [1] (::ReactiveMP.var"#3049#3050")(::Tuple{Symbol, ReactiveMP.Marginal{BayesBase.PointMass{Vector{Int64}}, Nothing}})
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:97
   [2] foreach(f::ReactiveMP.var"#3049#3050", itr::Base.Iterators.Zip{Tuple{Tuple{Symbol, Symbol}, Tuple{ReactiveMP.Marginal{BayesBase.PointMass{Vector{Int64}}, Nothing}, ReactiveMP.Marginal{BayesBase.PointMass{Matrix{Float64}}, Nothing}}}})
     @ Base ./abstractarray.jl:3187
   [3] discrete_transition_process_marginals
     @ ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:92 [inlined]
   [4] discrete_transition_structured_message_rule(message_names::Tuple{}, messages::Tuple{}, marginals_names::Tuple{Symbol, Symbol}, marginals::Tuple{ReactiveMP.Marginal{BayesBase.PointMass{Vector{Int64}}, Nothing}, ReactiveMP.Marginal{BayesBase.PointMass{Matrix{Float64}}, Nothing}}, q_a::ReactiveMP.Marginal{BayesBase.PointMass{Matrix{Float64}}, Nothing})
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:147
   [5] rule
     @ ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:184 [inlined]
   [6] MessageMapping
     @ ~/.julia/dev/ReactiveMP.jl/src/message.jl:353 [inlined]
   [7] as_message(message::ReactiveMP.DeferredMessage{Nothing, Tuple{ReactiveMP.MarginalObservable, ReactiveMP.MarginalObservable}, ReactiveMP.MessageMapping{ReactiveMP.DiscreteTransition, Val{:in}, ReactiveMP.Marginalisation, Nothing, Val{(:out, :a)}, Nothing, Nothing, Nothing, Nothing}}, cache::Nothing, messages::Nothing, marginals::Tuple{ReactiveMP.Marginal{BayesBase.PointMass{Vector{Int64}}, Nothing}, ReactiveMP.Marginal{BayesBase.PointMass{Matrix{Float64}}, Nothing}})
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/message.jl:235
   [8] as_message(message::ReactiveMP.DeferredMessage{Nothing, Tuple{ReactiveMP.MarginalObservable, ReactiveMP.MarginalObservable}, ReactiveMP.MessageMapping{ReactiveMP.DiscreteTransition, Val{:in}, ReactiveMP.Marginalisation, Nothing, Val{(:out, :a)}, Nothing, Nothing, Nothing, Nothing}}, cache::Nothing)
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/message.jl:231
   [9] as_message(message::ReactiveMP.DeferredMessage{Nothing, Tuple{ReactiveMP.MarginalObservable, ReactiveMP.MarginalObservable}, ReactiveMP.MessageMapping{ReactiveMP.DiscreteTransition, Val{:in}, ReactiveMP.Marginalisation, Nothing, Val{(:out, :a)}, Nothing, Nothing, Nothing, Nothing}})
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/message.jl:223
  [10] materialize!(type::ReactiveMP.EqualityRightOutbound, chain::ReactiveMP.EqualityChain{Rocket.ScheduleOnOperator{Rocket.AsapScheduler}, ReactiveMP.var"#15#17"{BayesBase.GenericProd, ReactiveMP.CompositeFormConstraint{Tuple{ReactiveMP.UnspecifiedFormConstraint, RxInfer.EnsureSupportedFunctionalForm}}}}, node_index::Int64)
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/nodes/equality.jl:125
  [11] (::ReactiveMP.ChainOutboundMapping)(::Tuple{Rocket.LazyObservable{Missing}, Rocket.SingleObservable{Missing, Rocket.AsapScheduler}})
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/nodes/equality.jl:165
```

RxInferExamples repo is fine and compiled without errors

**Comment by bvdmitri** [comment]
The problem is in unmatched `T`, first argument has `T=Float64` and the second has it `T=Int64`, I noticed this actually in tests that you changed [here](https://github.com/ReactiveBayes/ReactiveMP.jl/pull/455/files#diff-d0ffbcfb2e2f9f060973e3053c37646074e1da9280dbd81071089684f59f1ae6L66). I'm not sure if its 100% related. Maybe revert the test or add a new one with `Int` such that we catch this edge case in ReactiveMP repo as well without need to run RxInfer testset?

**Comment by wouterwln** [comment]
@bvdmitri should be fixed now

**Comment by bvdmitri** [comment]
The same test in RxInfer is still failing but for a different reason now 
```
ERROR: LoadError: DomainError with [NaN, NaN, NaN, NaN]:
Categorical: vector p is not a probability vector
Stacktrace:
   [1] #116
     @ ~/.julia/packages/Distributions/tQhJE/src/univariate/discrete/categorical.jl:30 [inlined]
   [2] check_args
     @ ~/.julia/packages/Distributions/tQhJE/src/utils.jl:89 [inlined]
   [3] #_#115
     @ ~/.julia/packages/Distributions/tQhJE/src/univariate/discrete/categorical.jl:30 [inlined]
   [4] DiscreteNonParametric
     @ ~/.julia/packages/Distributions/tQhJE/src/univariate/discrete/categorical.jl:29 [inlined]
   [5] (Distributions.Categorical{P} where P<:Real)(p::Vector{Float64})
     @ Distributions ~/.julia/packages/Distributions/tQhJE/src/univariate/discrete/categorical.jl:34
   [6] discrete_transition_structured_message_rule(message_names::Tuple{}, messages::Tuple{}, marginals_names::Tuple{Symbol, Symbol}, marginals::Tuple{ReactiveMP.Marginal{BayesBase.PointMass{Vector{Int64}}, Nothing}, ReactiveMP.Marginal{BayesBase.PointMass{Matrix{Float64}}, Nothing}}, q_a::ReactiveMP.Marginal{BayesBase.PointMass{Matrix{Float64}}, Nothing})
     @ ReactiveMP ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:166
   [7] rule
     @ ~/.julia/dev/ReactiveMP.jl/src/rules/discrete_transition/categoricals.jl:198 [inlined]
   [8] MessageMapping
     @ ~/.julia/dev/ReactiveMP.jl/src/message.jl:353 [inlined]
```
this is the only test that is failing

**Comment by wouterwln** [comment]
@bvdmitri it should pass now (when you use EF 2.0.3)

**Review by bvdmitri — APPROVED** [review]
